src/psyclone/psyGen.py

```diff
@@ -1768,15 +1800,21 @@ def gen_code(self, parent):
         for entity in self._children:
             entity.gen_code(parent)
 
-        if self.opencl:
-            # Ensure we block at the end of the invoke to ensure all
-            # kernels have completed before we return.
-            # This code ASSUMES only the first command queue is used for
-            # executing kernels.
+        if self.opencl and self._opencl_options['end_barrier']:
+
             parent.add(CommentGen(parent,
                                   " Block until all kernels have finished"))
-            parent.add(AssignGen(parent, lhs=flag,
-                                 rhs="clFinish(" + qlist + "(1))"))
+
+            # We need a clFinish for all the queues in the implementation
+            opencl_num_queues = 1
+            for kern in self.coded_kernels():
```

Comment:
This bit of code working out the number of queues is replicated from earlier. Could it be a method (or have the result stored somewhere)?

Reply:
They are not exactly the same as the one for the opencl_init gets the maximum of all Invokes to allocate the resources, while this one is the maximum for the specific Invoke because it just needs barriers for the queues it uses.

Is it true that the former could include a call to the latter, but it will still be two max() operations very similar to the current implementations, so I think it's better to keep it this way for now.

I also prefer not to have this value stored as even if it is not implemented yet I like the idea of some 'smart' scheduling transformation that changes the queues of each kernel at its convenience and it is only when we call `gen_code` that this has to be fully determined.